src/doc/trpl/unsafe.md

```diff
@@ -127,6 +111,101 @@ The latter assumption allows the compiler to optimize more
 effectively. As can be seen, actually *creating* a raw pointer is not
 unsafe, and neither is converting to an integer.
 
+## Reference and pointer aliasing
+
+Several examples can give a better comprehension of the reference and pointer aliasing rules.
+
+First of all, using `unsafe` code to incorrectly circumvent and violate the
+restrictions on references leads to undefined behaviour. For example,
+the following code creates two aliasing `&mut` references, and is therefore invalid.
+
+```
+use std::mem;
+let mut x: u8 = 1;
+
+let ref_1: &mut u8 = &mut x;
+let ref_2: &mut u8 = unsafe { mem::transmute(&mut *ref_1) };
+
+// oops, ref_1 and ref_2 point to the same piece of data (x) and are
+// both usable
+*ref_1 = 10;
+*ref_2 = 20;
+```
+
+When raw pointers come into play, the situation becomes more complex.
+Raw pointers can alias with anything, but at the same time they should respect
+the reference aliasing rules, so it's forbidden to write through a raw pointer,
+when the memory location it points to is borrowed by a safe Rust reference.
+
+```
+// This function prints 1, 2, 2 and one more number
+// The fourth number can be, for example, 2 in Release build and 3 in Debug build
```

Comment:
This doesn’t make it very clear the meaning of the program changes because the code invokes UB and compiler optimisation passes change the meaning.

> // The function prints 4 numbers. By violating the aliasing rules we will invoke undefined behaviour and the meaning of the program may change depending on the compilation options.

or something similar might be better. #brainstorming